File: robot_arm_node/commander/status_aggregator.py
#!/usr/bin/env python3
"""
@file   status_aggregator.py
@brief  状态聚合器 —— 订阅 /joint_states + TF2 → 合成 ArmStatus
@version 1.0
@date   2026-06-09

职责：
  - 维护当前关节位置/速度缓存
  - 通过 TF2 查询末端位姿（ArmPose）
  - 估算末端速度（ArmTwist，数值微分或从 joint_state.velocity 推算）
  - 构建 ArmStatus 消息（供 ArmCommanderNode 周期发布）
  - 提供查询接口：pose / joint_positions / is_moving / arm_at_target

被 ArmCommanderNode 持有，不独立运行。

@copyright Copyright (c) 2026 eMeet
"""

import math
import threading
from collections import deque

import rclpy
from rclpy.node import Node
from sensor_msgs.msg import JointState
from tf2_ros import TransformListener, Buffer

from robot_arm_interfaces.msg import ArmPose, ArmTwist, ArmStatus

from arm_utils import quat_to_rpy


# ── 常量 ──────────────────────────────────────────────────────────────────────────
JOINT_NAMES       = ['Joint1', 'Joint2', 'Joint3', 'Joint4', 'Joint5', 'Joint6']
EEF_LINK          = 'tool0'
BASE_FRAME        = 'arm_base_link'
JOINT_STATE_TOPIC = '/joint_states'

# 末端速度估算：保留最近 N 个位姿样本做数值微分
VELOCITY_WINDOW_SIZE = 10    # 100Hz 下 ≈0.1s 窗口
MIN_MOVING_VELOCITY  = 0.001  # m/s，低于此值认为静止


class StatusAggregator:
    """状态聚合器 —— 维护机械臂当前状态的权威来源。

    不继承 Node，共享 ArmCommanderNode 的 ROS 资源。
    提供线程安全的读写接口。
    """

    def __init__(self, node: Node):
        """
        Args:
            node: ArmCommanderNode 实例
        """
        self._node   = node
        self._logger = node.get_logger()

        # ── 关节状态缓存（线程安全）───────────────────────────────────────────
        self._joint_positions = {name: 0.0 for name in JOINT_NAMES}
        self._joint_velocities = {name: 0.0 for name in JOINT_NAMES}
        self._joint_lock = threading.Lock()

        node.create_subscription(
            JointState, JOINT_STATE_TOPIC,
            self._on_joint_state, 10,
        )

        # ── TF2 ───────────────────────────────────────────────────────────────
        self._tf_buffer   = Buffer()
        self._tf_listener = TransformListener(self._tf_buffer, node)

        # ── 位姿历史（用于速度估算）───────────────────────────────────────────
        self._pose_history = deque(maxlen=VELOCITY_WINDOW_SIZE)

        # ── 命令执行状态（由 ArmCommanderNode 更新）───────────────────────────
        self._executing_command_id = 0
        self._command_result       = ArmStatus.RESULT_NONE
        self._current_pose_state   = ArmStatus.POSE_STATE_OBSERVE  # 默认
        self._error_code           = ArmStatus.ERR_NONE
        self._is_moving            = False   # 由 Commander 状态机维护，避免速度微分抖动
        self._at_pose_start        = False   # 由运镜 server 维护：是否已到达运镜起始点
        self._camera_ready         = False   # 机械臂到达运镜起始点→运镜结束期间为 True
        self._is_tracking          = False   # IBVS 跟随激活期间为 True
        self._tracking_img_err     = 0.0     # 当前图像误差（归一化）
        self._tracking_depth_err_m = 0.0     # 当前深度误差（m）

        self._state_lock = threading.Lock()

        # ── 就绪 ──────────────────────────────────────────────────────────────
        self._logger.info('StatusAggregator 就绪')

    # ── 关节状态订阅 ──────────────────────────────────────────────────────────────
    def _on_joint_state(self, msg: JointState):
        now = self._node.get_clock().now().nanoseconds * 1e-9
        with self._joint_lock:
            for name, pos, vel in zip(msg.name, msg.position, msg.velocity):
                if name in self._joint_positions:
                    self._joint_positions[name] = pos
                    self._joint_velocities[name] = vel

        # 记录位姿历史（用于速度估算）
        pose = self._lookup_pose()
        if pose is not None:
            self._pose_history.append((now, pose))

    # ── 末端位姿查询 ──────────────────────────────────────────────────────────────
    @property
    def pose(self) -> ArmPose:
        """当前末端位姿（ArmPose 消息）。"""
        return self._lookup_pose() or self._zero_pose()

    def _lookup_pose(self) -> ArmPose | None:
        """从 TF2 查询末端位姿。"""
        try:
            t = self._tf_buffer.lookup_transform(
                BASE_FRAME, EEF_LINK, rclpy.time.Time())
            return self._transform_to_pose(t.transform)
        except Exception:
            return None

    def _zero_pose(self) -> ArmPose:
        p = ArmPose()
        p.x = p.y = p.z = 0.0
        p.roll = p.pitch = p.yaw = 0.0
        return p

    @staticmethod
    def _transform_to_pose(transform) -> ArmPose:
        """geometry_msgs/Transform → ArmPose。"""
        p = ArmPose()
        p.x = float(transform.translation.x)
        p.y = float(transform.translation.y)
        p.z = float(transform.translation.z)
        qx = transform.rotation.x
        qy = transform.rotation.y
        qz = transform.rotation.z
        qw = transform.rotation.w
        roll, pitch, yaw = quat_to_rpy(qx, qy, qz, qw)
        p.roll  = float(math.degrees(roll))
        p.pitch = float(math.degrees(pitch))
        p.yaw   = float(math.degrees(yaw))
        return p

    # ── 末端速度估算 ──────────────────────────────────────────────────────────────
    @property
    def twist(self) -> ArmTwist:
        """当前末端速度（数值微分估算，ArmTwist 消息）。"""
        if len(self._pose_history) < 2:
            return ArmTwist()   # 全零

        t0, p0 = self._pose_history[0]
        t1, p1 = self._pose_history[-1]
        dt = t1 - t0
        if dt < 1e-6:
            return ArmTwist()

        tw = ArmTwist()
        tw.vx     = float((p1.x - p0.x) / dt)
        tw.vy     = float((p1.y - p0.y) / dt)
        tw.vz     = float((p1.z - p0.z) / dt)
        tw.wroll  = float((p1.roll  - p0.roll)  / dt)
        tw.wpitch = float((p1.pitch - p0.pitch) / dt)
        tw.wyaw   = float((p1.yaw   - p0.yaw)   / dt)
        return tw

    # ── 关节状态查询 ──────────────────────────────────────────────────────────────
    @property
    def joint_positions(self) -> dict:
        """当前关节位置 {Joint1: rad, ...}。"""
        with self._joint_lock:
            return dict(self._joint_positions)

    @property
    def joint_velocities(self) -> dict:
        """当前关节速度 {Joint1: rad/s, ...}。"""
        with self._joint_lock:
            return dict(self._joint_velocities)

    def joint_position_list(self, names: list = None) -> list:
        """按指定顺序返回关节位置列表。"""
        if names is None:
            names = JOINT_NAMES
        with self._joint_lock:
            return [self._joint_positions[n] for n in names]

    # ── 运动状态 ──────────────────────────────────────────────────────────────────
    @property
    def is_moving(self) -> bool:
        with self._state_lock:
            return self._is_moving

    def set_moving(self, moving: bool):
        """由 Commander 状态机在 MOVING 进入/退出时调用，避免速度微分抖动。"""
        with self._state_lock:
            self._is_moving = moving

    @property
    def at_pose_start(self) -> bool:
        with self._state_lock:
            return self._at_pose_start

    def set_at_pose_start(self, at_start: bool):
        """由运镜 server 在到达起始点时置 True，新指令开始时（Commander）复位 False。"""
        with self._state_lock:
            self._at_pose_start = at_start

    @property
    def camera_ready(self) -> bool:
        with self._state_lock:
            return self._camera_ready

    def set_camera_ready(self, ready: bool):
        """到达运镜起始点时置 True，运镜结束（或新指令开始）时复位 False。"""
        with self._state_lock:
            self._camera_ready = ready

    def set_tracking(self, tracking: bool, img_err: float, depth_err_m: float):
        """由 TrackTargetServer 在跟随期间持续更新，退出时清零。"""
        with self._state_lock:
            self._is_tracking          = tracking
            self._tracking_img_err     = img_err
            self._tracking_depth_err_m = depth_err_m

    # ── 命令执行状态管理（由 ArmCommanderNode/action server 调用）─────────────────
    def set_command_state(self, command_id: int, result: int):
        """更新当前命令的执行状态。"""
        with self._state_lock:
            self._executing_command_id = command_id
            self._command_result       = result

    def set_pose_state(self, state: int):
        """更新当前姿态状态（STOWED / OBSERVE / SHOOTING）。"""
        with self._state_lock:
            self._current_pose_state = state

    def set_error(self, error_code: int):
        """设置错误码。"""
        with self._state_lock:
            self._error_code = error_code

    def get_error_code(self) -> int:
        """读取当前错误码。"""
        with self._state_lock:
            return self._error_code

    def clear_error(self):
        """清除错误码。"""
        with self._state_lock:
            self._error_code = ArmStatus.ERR_NONE

    # ── 构建 ArmStatus 消息 ───────────────────────────────────────────────────────
    def build_status_message(self) -> ArmStatus:
        """构建完整的 ArmStatus 消息（由 ArmCommanderNode 周期调用）。"""
        msg = ArmStatus()
        msg.header.frame_id = BASE_FRAME

        with self._state_lock:
            msg.current_pose_state  = self._current_pose_state
            msg.error_code          = self._error_code
            msg.executing_command_id = self._executing_command_id
            msg.command_result      = self._command_result

        msg.arm_pose     = self.pose
        msg.arm_twist    = self.twist
        msg.is_moving    = self.is_moving
        msg.arm_at_target = not self.is_moving   # TODO: 改为基于目标位姿的比较
        msg.arm_at_pose_start = self.at_pose_start
        msg.camera_ready      = self.camera_ready
        with self._state_lock:
            msg.is_tracking          = self._is_tracking
            msg.tracking_img_err     = float(self._tracking_img_err)
            msg.tracking_depth_err_m = float(self._tracking_depth_err_m)
        return msg

File: robot_arm_node/src/state/status_aggregator.cpp
/**
 * @file status_aggregator.cpp
 * @brief StatusAggregator 实现 —— 订阅 /joint_states + TF2 → ArmStatus
 *
 * on_joint_state 写关节缓存 + 追加位姿历史；pose() 每次 TF
 * lookup（失败返回零位姿）； twist()
 * 用位姿历史首末差分估速；build_status_message 汇总位姿/速度/flag。互斥保护，
 * TransformListener 传裸指针避 bad_weak_ptr（详见构造函数注释）。
 *
 * @version 1.0
 * @date 2026-07-01
 * @copyright Copyright (c) 2026 eMeet
 */
#include "robot_arm_node/state/status_aggregator.hpp"

#include <tf2/time.h>

#include <algorithm>
#include <cmath>

#include "robot_arm_node/motion/constants.hpp"
#include "robot_arm_node/motion/geometry.hpp"

namespace robot_arm_node::state {

using motion::BASE_FRAME;
using motion::EEF_LINK;
using motion::JOINT_NAMES;
using motion::JOINT_STATE_TOPIC;

/**
 * @brief 构造状态聚合器：订阅关节状态与控制模式，建立 TF2 监听。
 *
 * 关节缓存先按 JOINT_NAMES 全部填 0 —— 后续回调只更新「已存在的键」，
 * 这样 /joint_states 里混进的其他关节不会污染本臂的缓存。
 *
 * 控制模式话题用 transient_local：mode_manager_node 是 latched 广播，
 * 晚订阅也能立刻拿到当前模式，不必等下一次切换。
 *
 * TransformListener 传 &node_ 裸指针而不是 shared_from_this()：本对象常在
 * node 自身构造期内被创建，那时 weak_ptr 还没就绪，shared_from_this() 会抛
 * bad_weak_ptr。listener 只是瞬时借 node 的接口订阅 /tf，不持有所有权。
 *
 * @param node 宿主节点。
 */
StatusAggregator::StatusAggregator(rclcpp::Node& node)
    : node_(node), logger_(node.get_logger()) {
    // ──
    // 关节状态缓存初始化（全零）────────────────────────────────────────────────
    for (const auto& name : JOINT_NAMES) {
        joint_positions_[name] = 0.0;
        joint_velocities_[name] = 0.0;
    }

    joint_sub_ = node_.create_subscription<sensor_msgs::msg::JointState>(
        JOINT_STATE_TOPIC, 10, [this](const sensor_msgs::msg::JointState& msg) {
            this->on_joint_state(msg);
        });

    // 当前控制模式：mode_manager_node latched 广播，晚订阅也能立刻拿到当前值
    mode_sub_ = node_.create_subscription<ControlMode>(
        "/robot_arm/control_mode", rclcpp::QoS(1).transient_local().reliable(),
        [this](const ControlMode& msg) {
            std::lock_guard<std::mutex> lk(state_mtx_);
            active_control_mode_ = msg.mode;
        });

    // ── TF2
    // ──────────────────────────────────────────────────────────────────────
    tf_buffer_ = std::make_shared<tf2_ros::Buffer>(node_.get_clock());
    // 传裸指针而非 shared_from_this()：后者在「node
    // 自身构造期内创建本对象」时会抛 bad_weak_ptr。TransformListener 只是瞬时用
    // node 取接口订阅 /tf，不持有所有权。
    tf_listener_ =
        std::make_shared<tf2_ros::TransformListener>(*tf_buffer_, &node_);

    RCLCPP_INFO(logger_, "StatusAggregator 就绪");
}

/**
 * @brief 处理关节状态消息并记录末端位姿历史。
 * @param msg 收到的关节状态消息。
 */
// ── 关节状态订阅
// ────────────────────────────────────────────────────────────────
void StatusAggregator::on_joint_state(const sensor_msgs::msg::JointState& msg) {
    const double now = node_.get_clock()->now().seconds();

    {
        std::lock_guard<std::mutex> lk(joint_mtx_);
        const size_t m =
            std::min(msg.name.size(),
                     std::min(msg.position.size(), msg.velocity.size()));
        for (size_t i = 0; i < m; ++i) {
            auto it_p = joint_positions_.find(msg.name[i]);
            if (it_p != joint_positions_.end()) {
                it_p->second = msg.position[i];
                joint_velocities_[msg.name[i]] = msg.velocity[i];
            }
        }
    }

    // 记录位姿历史（用于速度估算）
    if (auto p = lookup_pose()) {
        std::lock_guard<std::mutex> lk(pose_mtx_);
        if (pose_history_.size() >= VELOCITY_WINDOW_SIZE)
            pose_history_.pop_front();
        pose_history_.emplace_back(now, std::move(*p));
    }
}

/**
 * @brief 获取当前末端位姿，查询失败时返回零位姿。
 * @return 当前末端位姿或默认构造的零位姿。
 */
// ── 末端位姿查询
// ────────────────────────────────────────────────────────────────
ArmPose StatusAggregator::pose() const {
    if (auto p = lookup_pose()) return *p;
    return ArmPose{};  // 零位姿（字段默认 0）
}

/**
 * @brief 通过 TF2 查询末端相对于基座的位姿。
 * @return 查询成功时返回位姿，否则返回空值。
 */
std::optional<ArmPose> StatusAggregator::lookup_pose() const {
    try {
        const auto t = tf_buffer_->lookupTransform(BASE_FRAME, EEF_LINK,
                                                   tf2::TimePointZero);
        return transform_to_pose(t.transform);
    } catch (const tf2::TransformException&) {
        return std::nullopt;
    }
}

/**
 * @brief 将 TF 变换转换为 ArmPose。
 * @param tf 待转换的几何变换。
 * @return 转换后的末端位姿，角度单位为度。
 */
ArmPose StatusAggregator::transform_to_pose(
    const geometry_msgs::msg::Transform& tf) {
    ArmPose p;
    p.x = static_cast<float>(tf.translation.x);
    p.y = static_cast<float>(tf.translation.y);
    p.z = static_cast<float>(tf.translation.z);
    const auto rpy = motion::quat_to_rpy(tf.rotation.x, tf.rotation.y,
                                         tf.rotation.z, tf.rotation.w);
    p.roll = static_cast<float>(rpy[0] * 180.0 / M_PI);
    p.pitch = static_cast<float>(rpy[1] * 180.0 / M_PI);
    p.yaw = static_cast<float>(rpy[2] * 180.0 / M_PI);
    return p;
}

/**
 * @brief 根据位姿历史首末样本估算末端速度。
 * @return 线速度和角速度；历史不足或时间间隔无效时返回零值。
 */
// ── 末端速度估算
// ────────────────────────────────────────────────────────────────
ArmTwist StatusAggregator::twist() const {
    std::lock_guard<std::mutex> lk(pose_mtx_);
    if (pose_history_.size() < 2) return ArmTwist{};

    const auto& [t0, p0] = pose_history_.front();
    const auto& [t1, p1] = pose_history_.back();
    const double dt = t1 - t0;
    if (dt < 1e-6) return ArmTwist{};

    ArmTwist tw;
    tw.vx = static_cast<float>((p1.x - p0.x) / dt);
    tw.vy = static_cast<float>((p1.y - p0.y) / dt);
    tw.vz = static_cast<float>((p1.z - p0.z) / dt);
    tw.wroll = static_cast<float>((p1.roll - p0.roll) / dt);
    tw.wpitch = static_cast<float>((p1.pitch - p0.pitch) / dt);
    tw.wyaw = static_cast<float>((p1.yaw - p0.yaw) / dt);
    return tw;
}

/**
 * @brief 获取全部缓存的关节位置。
 * @return 关节名称到位置的映射副本。
 */
// ── 关节状态查询
// ────────────────────────────────────────────────────────────────
std::map<std::string, double> StatusAggregator::joint_positions() const {
    std::lock_guard<std::mutex> lk(joint_mtx_);
    return joint_positions_;
}

/**
 * @brief 获取全部缓存的关节速度。
 * @return 关节名称到速度的映射副本。
 */
std::map<std::string, double> StatusAggregator::joint_velocities() const {
    std::lock_guard<std::mutex> lk(joint_mtx_);
    return joint_velocities_;
}

/**
 * @brief 按指定名称获取关节位置。
 * @param names 要查询的关节名称列表。
 * @return 与输入名称顺序对应的位置列表，未知关节返回零。
 */
std::vector<double> StatusAggregator::joint_position_list(
    const std::vector<std::string>& names) const {
    std::lock_guard<std::mutex> lk(joint_mtx_);
    std::vector<double> out;
    out.reserve(names.size());
    for (const auto& n : names) {
        auto it = joint_positions_.find(n);
        out.push_back(it != joint_positions_.end() ? it->second : 0.0);
    }
    return out;
}

/**
 * @brief 按指定名称获取关节速度。
 * @param names 要查询的关节名称列表。
 * @return 与输入名称顺序对应的速度列表（rad/s），未知关节返回零。
 */
std::vector<double> StatusAggregator::joint_velocity_list(
    const std::vector<std::string>& names) const {
    std::lock_guard<std::mutex> lk(joint_mtx_);
    std::vector<double> out;
    out.reserve(names.size());
    for (const auto& n : names) {
        auto it = joint_velocities_.find(n);
        out.push_back(it != joint_velocities_.end() ? it->second : 0.0);
    }
    return out;
}

/**
 * @brief 检查指定关节是否均存在于位置缓存中。
 * @param names 要检查的关节名称列表。
 * @return 名称列表非空且所有关节均存在时返回 true。
 */
bool StatusAggregator::has_joint_positions(
    const std::vector<std::string>& names) const {
    std::lock_guard<std::mutex> lk(joint_mtx_);
    for (const auto& n : names) {
        if (joint_positions_.find(n) == joint_positions_.end()) return false;
    }
    return !names.empty();
}

/**
 * @brief 获取当前控制模式。
 * @return 当前控制模式值。
 */
uint8_t StatusAggregator::active_control_mode() const {
    std::lock_guard<std::mutex> lk(state_mtx_);
    return active_control_mode_;
}

/**
 * @brief 获取机械臂运动状态。
 * @return 正在运动时返回 true。
 */
// ── 运动 / 运镜 / 跟随 flag
// ──────────────────────────────────────────────────────
bool StatusAggregator::is_moving() const {
    std::lock_guard<std::mutex> lk(state_mtx_);
    return is_moving_;
}

/**
 * @brief 设置机械臂运动状态。
 * @param moving 新的运动状态。
 */
void StatusAggregator::set_moving(bool moving) {
    std::lock_guard<std::mutex> lk(state_mtx_);
    is_moving_ = moving;
}

/** @brief 获取是否处于动作起始位姿。 @return 起始位姿状态。 */
bool StatusAggregator::at_pose_start() const {
    std::lock_guard<std::mutex> lk(state_mtx_);
    return at_pose_start_;
}

/** @brief 设置动作起始位姿状态。 @param at_start 新状态。 */
void StatusAggregator::set_at_pose_start(bool at_start) {
    std::lock_guard<std::mutex> lk(state_mtx_);
    at_pose_start_ = at_start;
}

/** @brief 获取相机就绪状态。 @return 相机就绪时返回 true。 */
bool StatusAggregator::camera_ready() const {
    std::lock_guard<std::mutex> lk(state_mtx_);
    return camera_ready_;
}

/** @brief 设置相机就绪状态。 @param ready 新的就绪状态。 */
void StatusAggregator::set_camera_ready(bool ready) {
    std::lock_guard<std::mutex> lk(state_mtx_);
    camera_ready_ = ready;
}

/**
 * @brief 更新视觉跟随状态及误差。
 * @param tracking 是否正在跟随。
 * @param img_err 图像误差。
 * @param depth_err_m 深度误差，单位为米。
 */
void StatusAggregator::set_tracking(bool tracking, double img_err,
                                    double depth_err_m) {
    std::lock_guard<std::mutex> lk(state_mtx_);
    is_tracking_ = tracking;
    tracking_img_err_ = img_err;
    tracking_depth_err_m_ = depth_err_m;
}

/**
 * @brief 更新当前命令的执行状态。
 * @param command_id 正在执行的命令标识。
 * @param result 命令执行结果。
 */
// ── 命令执行状态
// ────────────────────────────────────────────────────────────────
void StatusAggregator::set_command_state(uint32_t command_id, uint8_t result) {
    std::lock_guard<std::mutex> lk(state_mtx_);
    executing_command_id_ = command_id;
    command_result_ = result;
}

/** @brief 设置当前位姿状态。 @param state 新的位姿状态。 */
void StatusAggregator::set_pose_state(uint8_t state) {
    std::lock_guard<std::mutex> lk(state_mtx_);
    current_pose_state_ = state;
}

/** @brief 设置错误码。 @param error_code 新的错误码。 */
void StatusAggregator::set_error(uint8_t error_code) {
    std::lock_guard<std::mutex> lk(state_mtx_);
    error_code_ = error_code;
}

/** @brief 获取当前错误码。 @return 当前错误码。 */
uint8_t StatusAggregator::get_error_code() const {
    std::lock_guard<std::mutex> lk(state_mtx_);
    return error_code_;
}

/** @brief 清除当前错误并恢复为无错误状态。 */
void StatusAggregator::clear_error() {
    std::lock_guard<std::mutex> lk(state_mtx_);
    error_code_ = ArmStatus::ERR_NONE;
}

/**
 * @brief 汇总当前状态并构建 ArmStatus 消息。
 * @return 包含位姿、速度及各状态标志的状态消息。
 */
// ── 构建 ArmStatus 消息
// ─────────────────────────────────────────────────────────
ArmStatus StatusAggregator::build_status_message() const {
    ArmStatus msg;
    msg.header.frame_id = BASE_FRAME;

    // pose()/twist()/is_moving() 各自取锁，此处不能再持
    // state_mtx_（否则与它们死锁）
    msg.arm_pose = pose();
    msg.arm_twist = twist();

    std::lock_guard<std::mutex> lk(state_mtx_);
    msg.current_pose_state = current_pose_state_;
    msg.error_code = error_code_;
    msg.active_control_mode = active_control_mode_;
    msg.executing_command_id = executing_command_id_;
    msg.command_result = command_result_;
    msg.is_moving = is_moving_;
    msg.arm_at_target = !is_moving_;  // TODO: 改为基于目标位姿的比较
    msg.arm_at_pose_start = at_pose_start_;
    msg.camera_ready = camera_ready_;
    msg.is_tracking = is_tracking_;
    msg.tracking_img_err = static_cast<float>(tracking_img_err_);
    msg.tracking_depth_err_m = static_cast<float>(tracking_depth_err_m_);
    return msg;
}

}  // namespace robot_arm_node::state
